src/prometheus/definition.go

```diff
@@ -35,40 +40,54 @@ var FromRawEntityIDGenerator = func(_, rawEntityID string, _ definition.RawGroup
 // If group label is "container" then pod name is also included.
 func FromLabelValueEntityTypeGenerator(key string) definition.EntityTypeGeneratorFunc {
 	return func(groupLabel string, rawEntityID string, g definition.RawGroups, clusterName string) (string, error) {
-		switch groupLabel {
-		case "namespace", "node":
-			return fmt.Sprintf("k8s:%s:%s", clusterName, groupLabel), nil
-
-		case "container":
-			labels, err := getLabels(groupLabel, rawEntityID, key, g, "namespace", "pod")
-			if err != nil {
-				return "", err
-			}
-			if len(labels) != 2 {
-				return "", fmt.Errorf("cannot retrieve values for composing entity type for %q", groupLabel)
-			}
-			namespace := labels[0]
-			podName := labels[1]
-			if namespace == "" || podName == "" {
-				return "", fmt.Errorf("empty values for generated entity type for %q", groupLabel)
-			}
-			return fmt.Sprintf("k8s:%s:%s:%s:%s", clusterName, namespace, podName, groupLabel), nil
+		return defaultEntityTypeFromLabelValue(key, groupLabel, groupLabel, rawEntityID, g, clusterName)
+	}
+}
 
-		default:
-			labels, err := getLabels(groupLabel, rawEntityID, key, g, "namespace")
-			if err != nil {
-				return "", err
-			}
-			if len(labels) == 0 {
-				return "", fmt.Errorf("cannot retrieve values for composing entity type for %q", groupLabel)
-			}
-			namespace := labels[0]
+// FromLabelValueEntityTypeGeneratorWithCustomGroup generates the entity type in the same way
+// `FromLabelValueEntityTypeGenerator` does, but it uses the provided group instead of the group label to compose it.
+func FromLabelValueEntityTypeGeneratorWithCustomGroup(key string, group string) definition.EntityTypeGeneratorFunc {
+	return func(groupLabel string, rawEntityID string, g definition.RawGroups, clusterName string) (string, error) {
+		return defaultEntityTypeFromLabelValue(key, group, groupLabel, rawEntityID, g, clusterName)
+	}
+}
 
-			if namespace == "" {
-				return "", fmt.Errorf("empty namespace for generated entity type for %q", groupLabel)
-			}
-			return fmt.Sprintf("k8s:%s:%s:%s", clusterName, namespace, groupLabel), nil
+func defaultEntityTypeFromLabelValue( //nolint: cyclop
```

Comment:
What I find a bit confusing is that we have group and groupLabel. In one function those are the same since group and groupLabel are the same, in the second one they have different values.

Could we leverage an override concept?

Reply:
I wanted to define a really specific function for horizontalpodautoscaller entity type where it is called, but I couldn't do it because we would have circular dependencies.
On the other hand, some kind of override or having the caller passing both arguments would require changing all the current usages and again, I would like to avoid that kind of refactor in this PR so its scope doesn't get even wider.